subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java

```diff
@@ -302,7 +314,68 @@ public void visitMethodInsn(int opcode, String owner, String name, String descri
                     _INVOKESTATIC(className, INSTRUMENTED_CALL_SITE_METHOD, RETURN_CALL_SITE_ARRAY);
                     return;
                 }
+            } else if (opcode == Opcodes.INVOKEVIRTUAL && AbstractArchiveTaskCompatUtils.isArchiveTaskType(owner) && name.equals("setArchiveName") && descriptor.equals("(Ljava/lang/String;)V")) {
```

Comment:
I started refactoring `InstrumentingVisitor` in #15767 and introduced a `BackwardCompatibilityVisitor`. This visitor is already too messy.

Reply:
Thanks, @bamboo, I've seen that. I expected the overlap, but I needed to get started anyway since this problem is really complex. I'll wait for your changes to be merged, then I'll deal with the conflicts.

Reply:
FYI, I refactored the class visitor to a separate class. This way there's less coupling between the PRs.